### PR TITLE
Dagster integration and migration towards `pudl-models` repo

### LINKED ISSUE #66: Integrate Dagster to basic 10k/ex21 extraction
### Overview

The MVP infrastructure proposal (#63) for Sec10k extraction designates that we will use Dagster for orchestration. This will ensure the final design conforms to common tooling/design patterns we use in PUDL and elsewhere. This will also prepare the infrastructure here to be further integrated into PUDL in the future. For example, if we create a dagster deployment at some point in the future, it would potentially be able to control extraction from this repo.

Integrating dagster will also help to isolate various parts of the extraction process that don't always need to be run together. We can create a separate dag for training the ex21 model, run extraction on our validation sets and logging validation metrics, and running the full extraction.

### Success Criteria

How will we know that we're done?
* [ ] Settings handling is improved do CI does not fail intermittently
* [ ] Dag to run training
    * [ ] EX 21
* [ ] Dag to run extraction validation
    * [ ] Ex 21
    * [x] Basic 10k
* [ ] Dag to run full extraction
    * [ ] Ex 21
    * [ ] Basic 10k

### PR DESCRIPTION
**Pull request by zschira**
# Overview

Note: I tagged @jdangerx and @bendnorman, but probably just one infrastructure focused review is sufficient. Also this is a pretty huge PR, so I'm happy to run through it synchronously with anyone interested in reviewing.

Closes #66.

This PR integrates dagster into the mozilla sec extraction, as well as moving this repo towards becoming a generic `pudl-models` repo. To achieve this, I've reorganized this repo fairly significantly.

The new file structure looks like the following:

```
src/mozilla_sec_eia/library/
src/mozilla_sec_eia/models/
```

The `library/` directory contains generic tooling for `pudl-models`. This includes dagster resources/io-managers to enable easy integration with the mlflow tracking server, and methods to help constructing/configuring dagster jobs. These methods create 3 distinct 'types' of dagster jobs: production jobs, validation jobs, and training jobs. Validation and training jobs are executed with an in process dagster executor, which allows them to be backed by an mlflow run to allow full logging and tracking. Production are executed in a multi-process executor and are not associated with an mlflow run. These jobs can still use the mlflow api to load pre-trained models, but should not attempt to log metrics/parameters.

Each directory in `models/` is intended to be a standalone dagster code-location. This allows isolated development/execution of pudl models, and in the future could allow creating isolated python envs for each model making complex dependency management easier. To add a new model it needs to be added to `workspace.yaml`.  See the `README` for more info on execution and adding new models.

### CODE REVIEW

File: tests/integration/models/sec10k/extract_test.py
Comment (zschira) [inline tests/integration/models/sec10k/extract_test.py:26]: `xfail` due to weird errors on github runner from `xhtml2pdf`. Passing locally for me. Maybe @katie-lamb can try to run this test locally too?

Comment (zschira) [reply tests/integration/models/sec10k/extract_test.py:26]: Actually I think we might just be running out of memory on the runner. Gotta figure something out...

File: src/mozilla_sec_eia/models/sec10k/ex_21/inference.py
Comment (jdangerx) [inline src/mozilla_sec_eia/models/sec10k/ex_21/inference.py:237]: context: in my mind, this is what's happening - let me know what's wrong

* we generate a bunch of PDFs, then annotate them and munge that into a huggingface `Dataset` - afaik this is a unified interface that you can pass to any number of huggingface models?
* huggingface has a pretrained layoutlm model that we can just download
* we push the pdfs through the layoutlm model which gets us a dataframe with `id, subsidiary, loc, own_per` columns
* do some cleaning, tada!

File: src/mozilla_sec_eia/models/sec10k/ex_21/train_extractor.py
Comment (jdangerx) [inline src/mozilla_sec_eia/models/sec10k/ex_21/train_extractor.py:136]: non-blocking: Fine tuning is basically taking stock LayoutLM and then training it with our labeled data, so it knows what we're looking for?

File: src/mozilla_sec_eia/models/sec10k/utils/cloud.py
Comment (jdangerx) [inline src/mozilla_sec_eia/models/sec10k/utils/cloud.py:163]: hadn't really thought about it but the `ConfigurableResource` is much nicer than the `BaseSettings`, at least in a Dagster context

File: tests/conftest.py
Comment (jdangerx) [inline tests/conftest.py:65]: non-blocking: what do you think of making `_get_run` just a normal function that can get called, instead of involving it in the fixture machinery? the fixture machinery means you don't need to import it, I guess, but the readability/complexity cost of adding this as a factory function to work around fixture mechanics seems not worth it to me.

File: tests/integration/models/sec10k/extract_test.py
Comment (jdangerx) [reply tests/integration/models/sec10k/extract_test.py:26]: The PDF generation only runs 1 doc at a time, but we still run out of memory? That seems unexpected... did you end up figuring anything out here?

File: tests/unit/models/sec10k/utils_test.py
Comment (jdangerx) [inline tests/unit/models/sec10k/utils_test.py:35]: What's the test assertion here?

**Review by jdangerx — APPROVED** [review]
There's a lot here, so I definitely didn't catch everything. But the overall repo structure looks good to me and the dagster stuff doesn't seem too convoluted. I also think Ilearned a bunch about what was actually going on by reading this, although I still have plenty of questions - so thanks!

We should wait on @katie-lamb 's review of the more modelly bits before rushing to merge though.

**Comment by jdangerx** [comment]
Does look like you have some test failures though.

Beyond that we should talk about broader testing strategy here, but I think that can be a follow-up PR - this one's big enough already :)

File: README.rst
Comment (katie-lamb) [inline README.rst:95]: ``worspace.yaml`` typo

**Review by katie-lamb — APPROVED** [review]
This is a major structural change, so I'm sure there are some problems that will crop up, but I think we should get it merged. The `ex_21` job fails for me, I get the error: `pymupdf.EmptyFileError: Cannot open empty file: filename='/var/folders/c0/5zrbrqhx17d5jm6t03bw2nkw0000gn/T/tmppes1796q/897723-0000897723-20-000034.pdf'.
` (with different filenames each time I run it). Does it pass for you? Maybe I'm just getting a bad sample?

**Comment by zschira** [comment]
@katie-lamb I have seen the empty pdf issue crop up in some places while running the full extraction. I've got another branch where I'm collecting a bunch of error handling that comes up while getting through the extraction that I'll put in a follow up pr to this one.

File: src/mozilla_sec_eia/models/sec10k/utils/cloud.py
Comment (zschira) [reply src/mozilla_sec_eia/models/sec10k/utils/cloud.py:163]: Yeah they're super useful! No more passing around settings classes all over the place

File: README.rst
Comment (zschira) [reply README.rst]: Fixed!

File: tests/unit/models/sec10k/utils_test.py
Comment (zschira) [reply tests/unit/models/sec10k/utils_test.py:35]: This is just setting up a fixture that creates a dagster resource that gets used by other tests

File: tests/conftest.py
Comment (zschira) [reply tests/conftest.py]: I think that makes sense, changed now!

**Pull request merged**
